prusti-specs/src/rewriter.rs

```diff
@@ -209,6 +223,159 @@ impl AstRewriter {
+    pub fn process_closure_assertion(
+        &mut self,
+        spec_id: untyped::SpecificationId,
+        tokens: TokenStream,
+    ) -> syn::Result<syn::Item> {
+        let expr = Parser::from_token_stream(tokens).extract_assertion_expr()?;
+        let spec_id_str = spec_id.to_string();
+        let callsite_span = Span::call_site();
+        Ok(parse_quote_spanned! {callsite_span=>
+            #[allow(unused_must_use, unused_variables)]
+            {
+                #[prusti::spec_only]
+                #[prusti::spec_id = #spec_id_str]
+                || -> bool {
+                    #expr
+                };
+            }
+        })
+    }
+}
+
+pub fn translate_empty_assertion() -> syn::Expr {
+    parse_quote_spanned! {
+        Span::call_site() =>
+        true
+    }
+}
+
+pub fn translate_pledge_rhs_only(reference: Option<syn::Expr>, rhs: syn::Expr) -> syn::Expr {
+    if let Some(reference) = reference {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge_rhs(#reference, #rhs)
+        }
+    } else {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge_rhs(result, #rhs)
+        }
+    }
+}
+
+pub fn translate_pledge(reference: Option<syn::Expr>, lhs: syn::Expr, rhs: syn::Expr) -> syn::Expr {
+    if let Some(reference) = reference {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge(#reference, #lhs, #rhs)
+        }
+    } else {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge(result, #lhs, #rhs)
+        }
+    }
+}
+
+pub fn translate_implication(lhs: syn::Expr, rhs: syn::Expr) -> syn::Expr {
+    parse_quote_spanned! {
+        lhs.span().join(rhs.span()).unwrap() =>
+        implication(#lhs, #rhs)
+    }
+}
+
+pub fn translate_conjunction(mut conjuncts: Vec<syn::Expr>) -> syn::Expr {
+    debug_assert!(conjuncts.len() != 0, "empty conjuncts given");
+    if conjuncts.len() == 1 {
+        conjuncts.pop().unwrap()
+    } else {
+        let last = conjuncts.pop().unwrap();
+        let rest = translate_conjunction(conjuncts);
+        parse_quote_spanned! {
+            rest.span().join(last.span()).unwrap() =>
+            #rest && #last
+        }
+    }
+}
+
+fn args_to_tokens(mut args: Vec<(syn::Ident, syn::Type)>) -> TokenStream {
+    if args.len() == 0 {
+        TokenStream::new()
+    } else if args.len() == 1 {
+        let (ident, typ) = args.pop().unwrap();
+        quote_spanned! {
+            ident.span().join(typ.span()).unwrap() =>
+            #ident: #typ
+        }
+    } else {
+        let (ident, typ) = args.pop().unwrap();
+        let rest = args_to_tokens(args);
+        quote_spanned! {
+            rest.span().join(typ.span()).unwrap() =>
+            #rest, #ident: #typ
+        }
+    }
```

Comment:
Also `Iterator::reduce` (and `map`).

Reply:
Fixed in 6268917f30bbc875485f34bde92f5bd18cbde277